### ISSUE TITLE
Aggregating by Last or First

### ISSUE THREAD

**Issue by vijaydewan** [man-group/dtale#416]
Hi, I'm having issues when using the First & Last aggregation tools in Charts.  It doesn't seem to work properly for me.  Right now, I am getting different results when running these aggregations in charts than I am when I run the following code in Pandas: aggregation_results = chart_data.groupby(['cols'])[['agg_cols']].last().reset_index() (which when looking at charts/utils.py appears to be the same code used). Happy to email you a test file if you cannot replicate the results.

**Comment by aschonfeld** [comment]
@vjaydewan So the test I ran which produces different results in the chart versus running the code export was for data of the following shape:
| date | security_id |
| ------------- | ------------- |
| 20000101  | 1  |
| 20000101  | 2  |
| 20000101  | 3  |
| 20000102  | 1  |
| 20000102 | 2  |
| 20000102 | 3  |

And so on and so forth.  For the chart I choose `date` for the x-axis and `security_id` for the y-axis and then 'Keep First' as the aggregation.

Now my hypothesis as to why the data comes out differently between the two is because when using 'Keep First' & 'Keep Last' its not applying any sorting to the groups so I think the records are being chosen randomly.

So I think to get consistent results we would have to update the following line:

`chart_data = chart_data.groupby(['x'])[['security_id']].first().reset_index()`

to

```
chart_data = chart_data.groupby(['x'])[['security_id']].apply(
    lambda x: x.sort_values(by='security_id', ascending=True).head(1)
).reset_index()
```

Let me know what you think?  This will probably not be as performant as what we're doing now, but I doubt that inconsistent results are what we want either 🤣

**Comment by vijaydewan** [comment]
Hey Andrew, sorry for the delay! Thanks for looking into this. I think the less performant but more consistent solution make sense. It’s probably not going to be used too often so won’t affect overall performance much. I see you added a fix in the newest update. Can’t wait to check it out and thanks again!

**Comment by aschonfeld** [comment]
Sorry, would have included this with the release but was waiting for your response.  No worries, looks like i’ll be doing a hotfix to bring back python2.7 support so i’ll include it with that.

**Comment by aschonfeld** [comment]
Fixed in v1.33.1
